### PR TITLE
5.x tests on windows

### PR DESCRIPTION
**Pull request by n4bb12**
1. The PR adds windows to the CI matrix
1. The PR fixes some tests that were failing on windows by using a pattern instead of hardcoded strings.
1. Some tiny snapshots got converted to inline snapshots (imho makes it easier to understand tests)
1. [The lint job didn't run on GitHub's windows image due to line ending issues.](https://github.com/verdaccio/verdaccio/pull/1919/checks?check_run_id=1033411169#step:7:12) I fixed this by adding a [.gitattributes](https://github.com/alexkaratarakis/gitattributes/blob/master/Web.gitattributes) with `lf` as the default line ending for `text`
1. The PR sets one test on skip because it doesn't exit normally on windows, apparently due to a node.js streaming issue or something
1. Fixes `website/gatsby-node.js` on windows and simplifies it

### CODE REVIEW

**Comment by DanielRuf** [comment]
Awesome @n4bb12 :+1:

I guess you are still working on this and will add the missing strategies with the matrix setup.
Let us know if there is anything that we can help you with.

**Comment by n4bb12** [comment]
@DanielRuf this job is failing for me https://github.com/verdaccio/verdaccio/pull/1919/checks?check_run_id=1033477230
But it seems to already be misconfigured on the base branch since there is no lockfile in the repository. I don't think this is related to the PR changes.

**Comment by n4bb12** [comment]
I looked at the 5.x branch and the website jobs don't seem to run there. What am I missing?

**Comment by n4bb12** [comment]
It seems that there's an issue with pino not properly exiting on windows.

```
packages/verdaccio test: Test Suites: 1 passed, 1 total
packages/verdaccio test: Tests:       3 skipped, 152 passed, 155 total
packages/verdaccio test: Snapshots:   0 total
packages/verdaccio test: Time:        18.652 s
packages/verdaccio test: Ran all test suites matching /.\\test\\functional\\index*/i.
packages/verdaccio test:   ●  process.exit called with "1"
packages/verdaccio test:       at process.on._pino.default.final (../../logger/build/logger.js:162:13)
packages/verdaccio test:       at process.args (../../../node_modules/.pnpm/pino@5.17.0/node_modules/pino/lib/tools.js:354:12)
packages/verdaccio test: Failed
undefined
D:\a\verdaccio\verdaccio\packages\verdaccio:
 ERROR  verdaccio@5.0.0-alpha.0 test: `cross-env NODE_ENV=test jest --config ./test/jest.config.functional.js --testPathPattern ./test/functional/index* --passWithNoTests`
Exit status 1
 ERROR  Test failed. See above for more details.
##[error]Process completed with exit code 1.
```

**Comment by n4bb12** [comment]
```
packages/verdaccio test:  warn --- pino.final with prettyPrint does not support flushing
packages/verdaccio test:  fatal--- uncaughtException
```

[This happens somewhere in the middle of the test summary.](https://github.com/verdaccio/verdaccio/pull/1919/checks?check_run_id=1033587749#step:10:1937)

**Comment by n4bb12** [comment]
This might actually be happening due to verdaccio's use of pino.

**Comment by DanielRuf** [comment]
> But it seems to already be misconfigured on the base branch since there is no yarn lockfile in the repository. I don't think this is related to any of the PR changes.

Correct. This job is triggered by specific conditions.
https://github.com/verdaccio/verdaccio/blob/5.x/.github/workflows/ci-website.yml#L4-L10

I'm not sure if @juanpicado is already aware of this issue.

**Comment by DanielRuf** [comment]
> But it seems to already be misconfigured on the base branch since there is no yarn lockfile in the repository. I don't think this is related to any of the PR changes.

Correct. This job is triggered by specific conditions.
https://github.com/verdaccio/verdaccio/blob/5.x/.github/workflows/ci-website.yml#L4-L10

I'm not sure if @juanpicado is already aware of this issue. As far as I can see we moved to pnpmjs. So the `yarn` command is not used anymore. You can probably change this to the correct `pnpm` command.

**Comment by n4bb12** [comment]
> You can probably change this to the correct `pnpm` command.

I would prefer not to make yarn->pnpm in this PR.
Right now I'm only interested in getting tests to run on windows since it's impossible to contribute anything without being able to run tests.
I think the PR can be merged with the website jobs failing since they didn't break here.

**Comment by DanielRuf** [comment]
> It seems there's an issue with pino not properly exiting on windows.

Good catch. Not sure if this is an issue caused by pino. We use pino `5.17.0`.
https://github.com/verdaccio/verdaccio/blob/5.x/packages/logger/package.json#L27

The latest is pino 6.5.1:
https://github.com/pinojs/pino/releases

Maybe you can check the test which causes this.
Is the error always at this place?
```
packages/verdaccio test:       handling packages
packages/verdaccio test:         √ creating new package
packages/verdaccio test:         √ downloading non-existent tarball (3 ms)
packages/verdaccio test:  warn --- pino.final with prettyPrint does not support flushing
packages/verdaccio test:  fatal--- uncaughtException
packages/verdaccio test:         √ uploading incomplete tarball (22 ms)
```

I'm not yet sure what causes this on Windows but not on other platforms (and why).

**Comment by n4bb12** [comment]
Thanks for helping out.

Yes, it looks like it's always printed after the `downloading non-existent tarball` test.

If running just the `packages/verdaccio` tests where `downloading non-existent tarball` is included, it fails.
But if running just the `packages/auth` tests for example, it exits normally.

**Comment by n4bb12** [comment]
This single test fails: `should not store tarballs / chunked`

**Comment by n4bb12** [comment]
This single test passes but exits 1 on windows: `should not store tarballs / chunked`

**Comment by n4bb12** [comment]
After commenting this part in verdaccio's logger

```
  // process.on(
  //   'uncaughtException',
  //   pino.final(logger, (err, finalLogger) => {
  //     finalLogger.fatal(err, 'uncaughtException');
  //     process.exit(1);
  //   })
  // );
```

I get the following error that seems to have been swollowed:

```
Teardown Test Environment.
  console.error
    unhandledRejection Error: read ECONNRESET
        at TCP.onStreamRead (internal/stream_base_commons.js:208:20) {
      errno: -4077,
      code: 'ECONNRESET',
      syscall: 'read'
    }

      68 | 
      69 | process.on('unhandledRejection', function (err) {
    > 70 |   console.error('unhandledRejection', err);
         |           ^
      71 |   process.nextTick(function () {
      72 |     throw err;
      73 |   });

      at process.<anonymous> (functional/index.spec.ts:70:11)
```

**Comment by n4bb12** [comment]
Seems to be related https://github.com/nodejs/node/issues/27916#issuecomment-496025563

**Comment by n4bb12** [comment]
The "should not store tarballs" test is not typed and some of the code is years old. I don't know how to fix it and I don't really want to spend much more time on this. 😒

**Comment by juanpicado** [comment]
> After commenting this part in verdaccio's logger and rebuilding it
> 
> ```
>   // process.on(
>   //   'unhandledRejection',
>   //   pino.final(logger, (err, finalLogger) => {
>   //     finalLogger.fatal(err, 'uncaughtException');
>   //     process.exit(1);
>   //   })
>   // );
> ```
> 
> (btw, is it intentional that 'unhandledRejection' gets logged as 'uncaughtException' ?)
> 
> I get the following error that previously got swollowed by pino:
> 
> 2020-08-27 01_26_56-Hyper

Yes, it is. Refers to https://github.com/pinojs/pino/pull/528 

When I implemented Pino.js to replace the deprecated bunyan-node https://github.com/verdaccio/verdaccio/issues/1334 and https://github.com/verdaccio/verdaccio/pull/1760 I did not realize they don't recommend use prettifier for production environments and `pino.final` throw a warning https://github.com/pinojs/pino-pretty/issues/37 so I uncommented out until I fixed. It may deserve a comment for sure, my bad.

I'm not sure how to fix it, I'd need ideas, but, what I have on mind by far is start to use `env` parameters for prod and dev to verdaccio, and do not allow prettier in prod environments.

**Comment by juanpicado** [comment]
> the website job is failing for me

Let me fix that in a separated PR. I forgot update that last week.

**Comment by juanpicado** [comment]
> The "should not store tarballs" test is not typed and some of the code is years old. I don't know how to fix it and I don't really want to spend much more time on this. 😒

That unit test should be refactored and using new tooling we have already in place like `nock`. I've been removing those old test lately and make then more easy to maintain.  I'd vote for skip it, let's ad a comment and a `test.todo('xxx') so we don't forget.

**Comment by n4bb12** [comment]
> Yes, it is. Refers to [pinojs/pino#528](https://github.com/pinojs/pino/pull/528)

I wasn't referring to the pino warning, I was referring to 'unhandledRejection' gets logged as 'uncaughtException'.

**Review by juanpicado — APPROVED** [review]
Great work @n4bb12 !!! 🎉

**Comment by n4bb12** [comment]
@juanpicado is anything preventing this from being merged?

**Comment by juanpicado** [comment]
> @juanpicado is anything preventing this from being merged?

Actually was waiting for you :) it is your honor to merge.

**Comment by n4bb12** [comment]
github com_verdaccio_verdaccio_pull_1919

**Comment by DanielRuf** [comment]
Did you already accept the invitation? =)

Bildschirmfoto 2020-08-31 um 20 37 47

**Review by DanielRuf — APPROVED** [review]
LGTM, great work =) :+1:

**Comment by n4bb12** [comment]
> Did you already accept the invitation? =)
> 
> Bildschirmfoto 2020-08-31 um 20 37 47

I accepted the invitation and I can delete branches.

**Comment by juanpicado** [comment]
Actually you have rights to write.

**Comment by n4bb12** [comment]
I'm not re-creating the PR just to merge it myself. If you care about the PR, just one of you guys click on merge please.

**Comment by DanielRuf** [comment]
I'll check why the CI is stuck. Maybe this is the problem why you can not merge it. Because this can be only merged with administrator rights when the required checks are not fulfilled.

**Comment by DanielRuf** [comment]
Bildschirmfoto 2020-08-31 um 22 06 35
Bildschirmfoto 2020-08-31 um 22 06 11

Hm weird, is this resolved / fixed in the target branch?

File: .github/disabled/e2e-angular-cli-workflow.yml
Comment (DanielRuf) [inline .github/disabled/e2e-angular-cli-workflow.yml:46]: Isn't the matrix setup still missing here?

File: .github/disabled/e2e-gatsbyjs-cli-workflow.yml
Comment (DanielRuf) [inline .github/disabled/e2e-gatsbyjs-cli-workflow.yml:12]: Isn't the matrix setup still missing here?

File: .github/disabled/e2e-angular-cli-workflow.yml
Comment (juanpicado) [reply .github/disabled/e2e-angular-cli-workflow.yml:46]: Those are disabled, I moved there to don't block development, still need few changes to make then run with the new setup.

File: .gitattributes
Comment (DanielRuf) [inline .gitattributes:1]: Is this file needed for the tests? =)

Comment (n4bb12) [reply .gitattributes:1]: https://github.com/verdaccio/verdaccio/pull/1919#issue-474128576

File: .github/disabled/e2e-gatsbyjs-cli-workflow.yml
Comment (DanielRuf) [reply .github/disabled/e2e-gatsbyjs-cli-workflow.yml:12]: Nvm, seems the branch is updated now.

File: .github/disabled/e2e-angular-cli-workflow.yml
Comment (DanielRuf) [reply .github/disabled/e2e-angular-cli-workflow.yml:46]: Ah right. But normally I would prefer to add the matrix with such changes but it's ok for now =)

File: .gitattributes
Comment (DanielRuf) [reply .gitattributes:1]: Ah, thanks. I forgot about this.

Is the merge button activated for you now?

**Comment by juanpicado** [comment]
@DanielRuf @n4bb12 sorry the required checks for this branch were a bit mess up, now it is fine.

File: .github/disabled/e2e-gatsbyjs-cli-workflow.yml
Comment (n4bb12) [reply .github/disabled/e2e-gatsbyjs-cli-workflow.yml:12]: yes, I think so
I'll update the configs

**Comment by juanpicado** [comment]
> Bildschirmfoto 2020-08-31 um 22 06 35
> 
> Bildschirmfoto 2020-08-31 um 22 06 11
> 
> 
> 
> Hm weird, is this resolved / fixed in the target branch?

Code scanning is limited to master by now, since this branch still not published anywhere yet, can be re enabled anytime. Otherwise was blocking to merge.

File: .gitattributes
Comment (n4bb12) [reply .gitattributes:1]: github com_verdaccio_verdaccio_pull_1919

**Comment by n4bb12** [comment]
I'll fix the runs-on config mentioned by @DanielRuf

File: .gitattributes
Comment (DanielRuf) [reply .gitattributes:1]: Ok, then I'll merge this. Thanks for the feedback / screenshot.

**Comment by DanielRuf** [comment]
> I'll fix the runs-on config mentioned by @DanielRuf

Currently not needed, can be also done in a separate PR =)

Thank you very much for your contribution to verdaccio 🎉

**Comment by DanielRuf** [comment]
> I'll fix the runs-on config mentioned by @DanielRuf

Currently not needed, can be also done in a separate PR =)

Thank you very much for your contribution to verdaccio 👍 Welcome as collaborator 🎉

**Pull request merged**
